src/main/java/reactor/core/publisher/BaseSubscriber.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2011-2016 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import reactor.core.Exceptions;
+import reactor.core.Receiver;
+import reactor.core.Trackable;
+
+/**
+ * A simple base class for a {@link Subscriber} implementation that lets the user
+ * perform a {@link #request(long)} and {@link #cancel()} on it directly.
+ * <p>
+ * Override the {@link #doOnSubscribe(Subscription)}, {@link #doOnNext(Object)},
+ * {@link #doOnComplete()}, {@link #doOnError(Throwable)} and {@link #doOnCancel()} hooks
+ * to customize the base behavior. You also have a termination hook,
+ * {@link #doFinally(TerminationType)}.
+ * <p>
+ * Most of the time, exceptions triggered inside hooks are propagated to
+ * {@link #onError(Throwable)} (unless there is a fatal exception).
+ *
+ * @author Simon Baslé
+ */
+public class BaseSubscriber<T> implements Subscriber<T>, Subscription, Trackable, Receiver {
+
+	protected Subscription subscription;
+
+	@Override
+	public Subscription upstream() {
+		return subscription;
+	}
+
+	@Override
+	public boolean isStarted() {
+		return subscription != null;
+	}
+
+	/**
+	 * Hook for further processing of onSubscribe's Subscription.
+	 * Defaults to requesting {@code Long.MAX_VALUE}.
+	 *
+	 * @param subscription the subscription to optionally process
+	 */
+	protected void doOnSubscribe(Subscription subscription) {
```

Comment:
the prefix for the hooks that the user is supposed to override being `doOn` is probably confusing...